### PR TITLE
Allow users to specify custom mirroring and sync command

### PR DESCRIPTION
**Pull request by tulinkry**
Of course the change is compatible with the current setup:

mirror.conf:
```
commands:
  git: /usr/bin/git
```

The syntax is extended to support custom commands as follows:

```
commands:
  git:
    command: /usr/local/bin/git
    incoming: ['git', 'pull', '--dry-run']
    # for sync the default procedure will be used with the overriden command /usr/local/bin/git
```

```
commands:
  git:
    command: /usr/local/bin/git
    sync: ['git', 'pull', '--ff-only']
    # for mirroring the default procedure will be used with the overriden command /usr/local/bin/git
```

```
commands:
  git:
    command: /usr/local/bin/git
    incoming: ['git', 'pull', '--dry-run']
    sync: ['git', 'pull', '--ff-only']
```

This pull request is to allow user to overcome problems from #2583 and #2432 before an universal solution exists in opengrok.

### CODE REVIEW

**Comment by vladak** [comment]
Please do add copyrights.

File: opengrok-tools/src/test/python/test_mirror.py [hunk 1]
```diff
@@ -376,3 +398,183 @@ def mock_get_repo_type(*args):
...
+                         reason="requires /bin binaries")),
+        pytest.param('/usr/bin/echo',
+                     marks=pytest.mark.skipif(
+                         not os.path.exists('/usr/bin/echo'),
+                         reason="requires /usr/bin binaries")),
+    ]
+)
+def test_mirroring_custom_incoming_changes(echo_binary):
+    with tempfile.TemporaryDirectory() as repository_root:
+        repository = GitRepository(mock(), repository_root, 'test-1', {
```
Comment (vladak) [inline opengrok-tools/src/test/python/test_mirror.py:457]: do the tests assume that Git is installed on the system ?

File: opengrok-tools/src/test/python/test_mirror.py [hunk 2]
```diff
@@ -376,3 +398,183 @@ def mock_get_repo_type(*args):
...
+                         reason="requires /usr/bin binaries")),
+    ]
+)
+def test_mirroring_custom_sync_invoke_command(touch_binary):
+    with tempfile.TemporaryDirectory() as repository_root:
+        repository = GitRepository(mock(), repository_root, 'test-1', {
+            'sync': [touch_binary, 'sync.txt']
+        }, None, None, None)
+        assert repository.sync() == 0
+        assert 'sync.txt' in os.listdir(repository_root)
```
Comment (vladak) [inline opengrok-tools/src/test/python/test_mirror.py:531]: make 'sync.txt' a variable ?

File: opengrok-tools/src/test/python/test_mirror.py [hunk 3]
```diff
@@ -376,3 +398,183 @@ def mock_get_repo_type(*args):
...
+                         reason="requires /usr/bin binaries")),
+    ]
+)
+def test_mirroring_custom_incoming_invoke_command(touch_binary):
+    with tempfile.TemporaryDirectory() as repository_root:
+        repository = GitRepository(mock(), repository_root, 'test-1', {
+            'incoming': [touch_binary, 'incoming.txt']
+        }, None, None, None)
+        assert repository.incoming() is False
+        assert 'incoming.txt' in os.listdir(repository_root)
```
Comment (vladak) [inline opengrok-tools/src/test/python/test_mirror.py:440]: make 'incoming.txt' a variable ?

File: opengrok-tools/src/main/python/opengrok_tools/scm/repository.py [hunk 4]
```diff
@@ -59,6 +61,11 @@ def getCommand(self, cmd, **kwargs):
 
     def sync(self):
         # Eventually, there might be per-repository hooks added here.
+        if isinstance(self.configured_commands, dict) and \
```
Comment (vladak) [inline opengrok-tools/src/main/python/opengrok_tools/scm/repository.py:64]: refactor to a function and use that in incoming() as well ?

**Comment by vladak** [comment]
I will update https://github.com/oracle/opengrok/wiki/Repository-synchronization once this is in.

File: opengrok-tools/src/test/python/test_mirror.py (hunk 1 above)
Comment (tulinkry) [reply opengrok-tools/src/test/python/test_mirror.py:457]: No no, the git command is never executed, most of the tests uses custom commands (like echo or true/false) and the rest uses mocking for the original call.

File: opengrok-tools/src/main/python/opengrok_tools/scm/repository.py (hunk 4 above)
Comment (tulinkry) [reply opengrok-tools/src/main/python/opengrok_tools/scm/repository.py]: done

File: opengrok-tools/src/test/python/test_mirror.py (hunk 3 above)
Comment (tulinkry) [reply opengrok-tools/src/test/python/test_mirror.py]: done

File: opengrok-tools/src/test/python/test_mirror.py (hunk 2 above)
Comment (tulinkry) [reply opengrok-tools/src/test/python/test_mirror.py]: done

**Comment by tulinkry** [comment]
Copyrights updated, I can write the portion for the wiki so you just can copypaste it

**Comment by vladak** [comment]
> Copyrights updated, I can write the portion for the wiki so you just can copypaste it

That would be nice.

File: opengrok-tools/src/main/python/opengrok_tools/scm/repository.py [hunk 5]
```diff
@@ -18,10 +18,12 @@
 #
 
 #
-# Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2018, 2020, Oracle and/or its affiliates. All rights reserved.
```
Comment (vladak) [inline opengrok-tools/src/main/python/opengrok_tools/scm/repository.py:21]: this one don't change

Comment (tulinkry) [reply opengrok-tools/src/main/python/opengrok_tools/scm/repository.py]: changed

File: opengrok-tools/src/main/python/opengrok_tools/scm/repository.py [hunk 6]
```diff
@@ -59,6 +62,10 @@ def getCommand(self, cmd, **kwargs):
 
     def sync(self):
         # Eventually, there might be per-repository hooks added here.
+        if self.is_command_overridden(self.configured_commands, 'sync'):
```
Comment (vladak) [inline opengrok-tools/src/main/python/opengrok_tools/scm/repository.py:65]: the 'sync' string could be a variable

File: opengrok-tools/src/main/python/opengrok_tools/scm/repository.py [hunk 7]
```diff
@@ -75,6 +82,105 @@ def incoming(self):
         """
         Check if there are any incoming changes.
 
+        Return True if so, False otherwise.
+        """
+        if self.is_command_overridden(self.configured_commands, 'incoming'):
```
Comment (vladak) [inline opengrok-tools/src/main/python/opengrok_tools/scm/repository.py:87]: ditto

File: opengrok-tools/src/main/python/opengrok_tools/scm/repository.py [hunk 8]
```diff
@@ -75,6 +82,105 @@ def incoming(self):
...
+
+    def incoming_check(self):
+        """
+        Check if there are any incoming changes.
+
         Return True if so, False otherwise.
         """
         return True
+
+    def _run_custom_sync_command(self, command):
```
Comment (vladak) [inline opengrok-tools/src/main/python/opengrok_tools/scm/repository.py:101]: is this something that could be reused in the individual repository implementations ? same for _run_custom_incoming_command().

File: opengrok-tools/src/main/python/opengrok_tools/scm/repository.py [hunk 9]
```diff
@@ -75,6 +82,105 @@ def incoming(self):
...
+        :param command: the command
+        :return: tuple of (status, output)
+                    - status: 0 on success execution, non-zero otherwise
+                    - output: command output as string
+        """
+        cmd = self.getCommand(command, work_dir=self.path,
+                              env_vars=self.env, logger=self.logger)
+        cmd.execute()
+        if cmd.getretcode() != 0 or cmd.getstate() != Command.FINISHED:
+            cmd.log_error("failed to perform command")
```
Comment (vladak) [inline opengrok-tools/src/main/python/opengrok_tools/scm/repository.py:143]: log the actual command

Comment (vladak) [reply opengrok-tools/src/main/python/opengrok_tools/scm/repository.py:143]: and the retcode + state

File: opengrok-tools/src/test/python/test_mirror.py [hunk 10]
```diff
@@ -376,3 +400,185 @@ def mock_get_repo_type(*args):
             repos = get_repos_for_project(project_name, None, source_root,
                                           ignored_repos=['.'])
             assert len(repos) == 0
+
+
+@pytest.mark.parametrize(['config', 'expected_command'], [
```
Comment (vladak) [inline opengrok-tools/src/test/python/test_mirror.py:405]: expected first, actual second ?

File: opengrok-tools/src/test/python/test_mirror.py [hunk 11]
```diff
@@ -376,3 +400,185 @@ def mock_get_repo_type(*args):
...
+    (None, 'default-command'),
```
Comment (vladak) [inline opengrok-tools/src/test/python/test_mirror.py:406]: make 'default-command' a variable ?

File: opengrok-tools/src/main/python/opengrok_tools/scm/repository.py [hunk 12]
```diff
@@ -75,6 +82,105 @@ def incoming(self):
...
+                cmd.geterroutputstr()
+            ]))
+        return 0, cmd.getoutputstr()
+
+    @staticmethod
+    def _repository_command(configured_commands, default=lambda: None):
+        """
+        Get the repository command, or use default supplier.
+
+        :param configured_commands: commands section from configuration
```
Comment (vladak) [inline opengrok-tools/src/main/python/opengrok_tools/scm/repository.py:158]: this should probably say that it looks up the 'command' key

File: opengrok-tools/src/main/python/opengrok_tools/scm/repository.py (hunk 9 above)
Comment (tulinkry) [reply opengrok-tools/src/main/python/opengrok_tools/scm/repository.py:143]: that's part of the method
```
    def log_error(self, msg):
        if self.state is Command.FINISHED:
            self.logger.error("{}: command {} in directory {} exited with {}".
                              format(msg, self.cmd, self.work_dir,
                                     self.getretcode()))
        else:
            self.logger.error("{}: command {} in directory {} ended with "
                              "invalid state {}".
                              format(msg, self.cmd, self.work_dir, self.state))

File: opengrok-tools/src/main/python/opengrok_tools/scm/repository.py (hunk 12 above)
Comment (tulinkry) [reply opengrok-tools/src/main/python/opengrok_tools/scm/repository.py:158]: when saying that I should probably mention that it does lookup only when configuration is a dict and that is exactly what is written in the code...so I'm not a big fan rephrasing the code to the comment

Comment (tulinkry) [reply opengrok-tools/src/main/python/opengrok_tools/scm/repository.py:158]: When I add that, I should perhaps mention that it only does the lookup when the configured_commands is a dictionary and that all together....is exactly what is in the code. So I'm not a big fan adding that comment.

File: opengrok-tools/src/test/python/test_mirror.py (hunk 10 above)
Comment (tulinkry) [reply opengrok-tools/src/test/python/test_mirror.py:405]: done

File: opengrok-tools/src/test/python/test_mirror.py (hunk 11 above)
Comment (tulinkry) [reply opengrok-tools/src/test/python/test_mirror.py:406]: done

File: opengrok-tools/src/main/python/opengrok_tools/scm/repository.py (hunk 8 above)
Comment (tulinkry) [reply opengrok-tools/src/main/python/opengrok_tools/scm/repository.py:104]: I think it's possible, didn't do it for mercurial as it contains more complicated logic.

Comment (tulinkry) [reply opengrok-tools/src/main/python/opengrok_tools/scm/repository.py:104]: I split it to a commit per repository type, could you review it again if that is ok? I don't have enough resources to test all of the repository types.

**Comment by vladak** [comment]
Thanks !

**Pull request merged**